### ISSUE TITLE
Project level default reviewers cause events not to be sent for updates

### ISSUE THREAD

**Issue by mkobit** [Eernie/bitbucket-webhooks-plugin#52]
In Bitbucket 5.1, default reviewers for project level was added (see https://confluence.atlassian.com/bitbucketserver/bitbucket-server-5-1-release-notes-902140366.html). It doesn't look like updates are sent for repositories that have this enabled.

1. Configure default reviewers at the project level
1. Setup a repository
1. Open a pull request
1. Push new changes to the same "from" of the pull request
1. No event is triggered for update

Stacktrace:

```
2017-07-25 15:56:41,728 WARN  [AtlassianEvent::thread-11]  c.a.s.i.h.r.DefaultRepositoryHookService [AUTH/auth-service[2403]] Error calling com.atlassian.bitbucket.internal.defaultreviewers.mergecheck.RequiredReviewersMergeCheck.preUpdate (com.atlassian.bitbucket.server.bitbucket-default-reviewers:required-reviewers-merge-check)
com.atlassian.bitbucket.AuthorisationException: You are not permitted to access this resource
    at com.atlassian.stash.internal.aop.ExceptionRewriteAdvice.afterThrowing(ExceptionRewriteAdvice.java:35)
    at com.atlassian.plugin.util.ContextClassLoaderSettingInvocationHandler.invoke(ContextClassLoaderSettingInvocationHandler.java:26)
    at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:56)
    at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:60)
    at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:70)
    at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:53)
    at org.eclipse.gemini.blueprint.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:57)
    at com.atlassian.bitbucket.internal.defaultreviewers.DefaultPullRequestConditionService.getScope(DefaultPullRequestConditionService.java:263)
    at com.atlassian.bitbucket.internal.defaultreviewers.DefaultPullRequestConditionService.toPullRequestCondition(DefaultPullRequestConditionService.java:295)
    at com.atlassian.bitbucket.internal.defaultreviewers.DefaultPullRequestConditionService.lambda$getPullRequestConditions$0(DefaultPullRequestConditionService.java:121)
    at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
    at java.util.Iterator.forEachRemaining(Iterator.java:116)
    at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
    at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
    at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
    at com.atlassian.bitbucket.util.PageImpl.transform(PageImpl.java:137)
    at com.atlassian.bitbucket.util.PageImpl.transform(PageImpl.java:15)
    at com.atlassian.bitbucket.internal.defaultreviewers.DefaultPullRequestConditionService.getPullRequestConditions(DefaultPullRequestConditionService.java:121)
    at com.atlassian.bitbucket.internal.defaultreviewers.mergecheck.RequiredReviewersMergeCheck.lambda$preUpdate$0(RequiredReviewersMergeCheck.java:64)
    at com.atlassian.bitbucket.util.PagedIterable.iterator(PagedIterable.java:27)
    at java.lang.Iterable.spliterator(Iterable.java:101)
    at com.atlassian.bitbucket.util.MoreStreams.streamIterable(MoreStreams.java:86)
    at com.atlassian.bitbucket.internal.defaultreviewers.mergecheck.RequiredReviewersMergeCheck.preUpdate(RequiredReviewersMergeCheck.java:75)
    at com.atlassian.bitbucket.internal.defaultreviewers.mergecheck.RequiredReviewersMergeCheck.preUpdate(RequiredReviewersMergeCheck.java:36)
    at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:583)
    at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:284)
    at com.atlassian.stash.internal.pull.DefaultMergeRequestCheckService.doCheck(DefaultMergeRequestCheckService.java:61)
    at com.atlassian.stash.internal.pull.DefaultMergeRequestCheckService.checkMergeability(DefaultMergeRequestCheckService.java:45)
    at com.atlassian.stash.internal.pull.DefaultPullRequestService.canMerge(DefaultPullRequestService.java:246)
    at com.atlassian.plugin.util.ContextClassLoaderSettingInvocationHandler.invoke(ContextClassLoaderSettingInvocationHandler.java:26)
    at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:56)
    at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:60)
    at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:70)
    at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:53)
    at org.eclipse.gemini.blueprint.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:57)
    at nl.topicus.bitbucket.api.PullRequestListener.rescopedEvent(PullRequestListener.java:94)
    at com.atlassian.event.internal.SingleParameterMethodListenerInvoker.invoke(SingleParameterMethodListenerInvoker.java:40)
    at com.atlassian.stash.internal.event.AsyncBatchingInvokersTransformer$AsyncInvokerBatch.invoke(AsyncBatchingInvokersTransformer.java:109)
    at com.atlassian.event.internal.AsynchronousAbleEventDispatcher$1$1.run(AsynchronousAbleEventDispatcher.java:38)
    at com.atlassian.sal.core.executor.ThreadLocalDelegateRunnable.run(ThreadLocalDelegateRunnable.java:34)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.lang.Thread.run(Thread.java:745)
    ... 129 frames trimmed
Caused by: org.springframework.security.access.AccessDeniedException: Access is denied
    at org.springframework.security.access.expression.method.ExpressionBasedPostInvocationAdvice.after(ExpressionBasedPostInvocationAdvice.java:77)
    ... 45 common frames omitted
2017-07-25 15:56:41,767 ERROR [AtlassianEvent::thread-11]  c.a.s.i.e.AsyncBatchingInvokersTransformer There was an exception thrown trying to dispatch event 'com.atlassian.bitbucket.event.pull.PullRequestRescopedEvent[source=com.atlassian.stash.internal.pull.DefaultPullRequestService@61cf2cf7]' for the invoker 'SingleParameterMethodListenerInvoker{method=public void nl.topicus.bitbucket.api.PullRequestListener.rescopedEvent(com.atlassian.bitbucket.event.pull.PullRequestRescopedEvent) throws java.io.IOException, listener=nl.topicus.bitbucket.api.PullRequestListener@16291ded}'
java.lang.RuntimeException: Transaction rolled back because it has been marked as rollback-only. Listener: nl.topicus.bitbucket.api.PullRequestListener event: com.atlassian.bitbucket.event.pull.PullRequestRescopedEvent
    at com.atlassian.event.internal.SingleParameterMethodListenerInvoker.invoke(SingleParameterMethodListenerInvoker.java:55)
    at com.atlassian.stash.internal.event.AsyncBatchingInvokersTransformer$AsyncInvokerBatch.invoke(AsyncBatchingInvokersTransformer.java:109)
    at com.atlassian.event.internal.AsynchronousAbleEventDispatcher$1$1.run(AsynchronousAbleEventDispatcher.java:38)
    at com.atlassian.sal.core.executor.ThreadLocalDelegateRunnable.run(ThreadLocalDelegateRunnable.java:34)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.lang.Thread.run(Thread.java:745)
    ... 1 frame trimmed
Caused by: org.springframework.transaction.UnexpectedRollbackException: Transaction rolled back because it has been marked as rollback-only
    at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:724)
    at com.atlassian.plugin.util.ContextClassLoaderSettingInvocationHandler.invoke(ContextClassLoaderSettingInvocationHandler.java:26)
    at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:56)
    at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:60)
    at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:70)
    at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:53)
    at org.eclipse.gemini.blueprint.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:57)
    at nl.topicus.bitbucket.api.PullRequestListener.rescopedEvent(PullRequestListener.java:94)
    at com.atlassian.event.internal.SingleParameterMethodListenerInvoker.invoke(SingleParameterMethodListenerInvoker.java:40)
    ... 6 common frames omitted

```

Note the `Error calling com.atlassian.bitbucket.internal.defaultreviewers.mergecheck.RequiredReviewersMergeCheck.preUpdate (com.atlassian.bitbucket.server.bitbucket-default-reviewers:required-reviewers-merge-check)`

I'm not sure if this is a Bitbucket Server issue or a plugin issue.

**Comment by mkobit** [comment]
I don't know if this is fixed by https://confluence.atlassian.com/bitbucketserver/bitbucket-server-5-2-release-notes-917966093.html either.

**Comment by bturner** [comment]
@mkobit It's not fixed by 5.2.0, no. However, I've created [BSERV-10045](https://jira.atlassian.com/browse/BSERV-10045) in relation to this issue. The `LICENSED_USER` check being performed when `PullRequestService.canMerge` is called is incorrect, and should be removed.

For compatibility reasons, this add-on may need to be updated to perform some escalation (`REPO_READ` + `LICENSED_USER` would be enough), at least until the minimum supported Bitbucket Server version is 5.4.0 or higher, just to ensure it behaves consistently on versions with and without the BSERV-10045 fix.  I just wanted to note that I believe the real issue here is in Bitbucket Server, and we'll be addressing it.

For administrators of instances affected by this issue, you can watch BSERV-10045 to be notified which versions you can upgrade to to avoid the error.

**Comment by casz** [comment]
@bturner this should be avoided on plugin side with the use of direct scm `canMerge`/`tryMerge` in 1.5.0?

**Comment by bturner** [comment]
Correct. Now that merge checks are not actually run, they can't cause failures anymore.

**Comment by casz** [comment]
Then I am closing the issue 👍 as Fixed in 1.5.0
